File: playwright-python/steps/cart_steps.py
from decimal import Decimal

from pytest_bdd import given, then, when

from clients.storefront_api import StorefrontApi
from pages.cart_page import CartPage


@given("the cart has been cleared")
def cart_has_been_cleared(storefront_api: StorefrontApi) -> None:
    response = storefront_api.delete("/cart")
    assert response.ok, response.text()


@given("the catalog contains an available product")
def catalog_contains_available_product(cart_products: dict[str, object]) -> None:
    assert cart_products["primary"].in_stock > 0


@given("the customer opens the storefront")
def customer_opens_storefront(cart_page: CartPage) -> None:
    cart_page.open()


@then("the floating cart is not visible")
@then("the floating cart is hidden")
def floating_cart_is_not_visible(cart_page: CartPage) -> None:
    cart_page.expect_cart_hidden()


@when("the customer adds 1 available product to the cart")
def customer_adds_one_available_product(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.add_product(cart_products["primary"].name, 1)


@then("the floating cart is visible")
def floating_cart_is_visible(cart_page: CartPage) -> None:
    cart_page.expect_cart_visible()


@then("it shows the product name")
def cart_shows_product_name(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.expect_product_in_cart(cart_products["primary"].name)


@then("it shows 1 total item")
def cart_shows_one_total_item(cart_page: CartPage) -> None:
    cart_page.expect_total_items(1)


@then("it shows the correct total price")
def cart_shows_correct_total(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.expect_grand_total(cart_products["primary"].price)


@given("the customer has added a product to the cart")
@given("the customer has added 1 product to the cart")
def customer_has_added_product(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.add_product(cart_products["primary"].name, 1)


@when("the customer scrolls to another part of the storefront")
def customer_scrolls_storefront(cart_page: CartPage) -> None:
    cart_page.scroll_to_top()


@then("the floating cart remains inside the viewport")
def floating_cart_remains_in_viewport(cart_page: CartPage) -> None:
    cart_page.expect_cart_inside_viewport()


@when("the customer selects quantity 3")
def customer_selects_quantity_three(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.select_quantity(cart_products["primary"].name, 3)


@when("adds the product to the cart")
def adds_selected_product(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.add_selected_product(cart_products["primary"].name)


@then("the cart line shows quantity 3")
@then("the line quantity is 3")
def cart_line_shows_quantity_three(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.expect_line_quantity(cart_products["primary"].name, 3)


@then("the total equals three times the unit price")
def total_equals_three_times_price(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.expect_grand_total(cart_products["primary"].price * 3)


@when("the customer adds 2 more of the same product")
def customer_adds_two_more(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.add_product(cart_products["primary"].name, 2)


@then("the cart contains one line for that product")
def cart_contains_one_line(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.expect_one_line(cart_products["primary"].name)


def _add_multiple_products(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.add_product(cart_products["primary"].name, 1)
    cart_page.add_product(cart_products["secondary"].name, 2)


@when("the customer adds multiple different products")
def customer_adds_multiple_products(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    _add_multiple_products(cart_page, cart_products)


@then("every selected product is shown in the cart")
def every_product_is_shown(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.expect_product_in_cart(cart_products["primary"].name)
    cart_page.expect_product_in_cart(cart_products["secondary"].name)


@then("total items equal the sum of all quantities")
def total_items_equal_quantity_sum(cart_page: CartPage) -> None:
    cart_page.expect_total_items(3)


@then("the grand total equals the sum of all line totals")
def grand_total_equals_line_totals(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    expected = cart_products["primary"].price + cart_products["secondary"].price * 2
    cart_page.expect_grand_total(expected)


@given("the cart contains two different products")
def cart_contains_two_products(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    _add_multiple_products(cart_page, cart_products)


@when("the customer removes one product")
def customer_removes_one_product(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.remove_product(cart_products["primary"].name)


@then("only that product disappears from the cart")
def only_removed_product_disappears(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.expect_product_not_in_cart(cart_products["primary"].name)
    cart_page.expect_product_in_cart(cart_products["secondary"].name)


@then("the totals are recalculated")
def totals_are_recalculated(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.expect_total_items(2)
    cart_page.expect_grand_total(cart_products["secondary"].price * 2)


@given("the cart contains products")
def cart_contains_products(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    _add_multiple_products(cart_page, cart_products)


@when('the customer selects "Clear"')
def customer_selects_clear(cart_page: CartPage) -> None:
    cart_page.clear()


@then("all cart lines are removed")
def all_cart_lines_are_removed(cart_page: CartPage) -> None:
    assert cart_page.page.locator("#cart-items .cart-item").count() == 0


@when("the customer reloads the storefront")
def customer_reloads_storefront(cart_page: CartPage) -> None:
    cart_page.reload()


@then("the same product and quantity remain in the cart")
def same_product_and_quantity_remain(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.expect_product_in_cart(cart_products["primary"].name)
    cart_page.expect_line_quantity(cart_products["primary"].name, 1)


@when("the customer attempts to add more units than available")
def customer_attempts_over_stock(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    product = cart_products["primary"]
    cart_page.attempt_to_add_over_stock(product.name, product.in_stock)


@then("the request is rejected")
def request_is_rejected(cart_page: CartPage) -> None:
    cart_page.expect_request_rejected()


@then("a stock error is displayed")
def stock_error_is_displayed(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    product = cart_products["primary"]
    cart_page.expect_stock_error(product.in_stock, product.name)


@then("the cart quantity is unchanged")
def cart_quantity_is_unchanged(cart_page: CartPage) -> None:
    cart_page.expect_total_items(0)


@when("the customer removes that product from the cart")
def customer_removes_added_product(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    cart_page.remove_product(cart_products["primary"].name)


@then("the catalog shows the original available stock")
def catalog_shows_original_stock(
    cart_page: CartPage, cart_products: dict[str, object]
) -> None:
    product = cart_products["primary"]
    cart_page.expect_catalog_stock(product.name, product.in_stock)

File: playwright-typescript/src/steps/cart-steps.ts
import assert from "node:assert/strict";
import { randomUUID } from "node:crypto";
import { Given, Then, When } from "@cucumber/cucumber";
import { CartPage } from "../pages/cart-page.js";
import type { FictoshopWorld } from "../support/world.js";

function cart(world: FictoshopWorld): CartPage { return new CartPage(world.page); }
function primary(world: FictoshopWorld) { return world.state.products.primary; }
function secondary(world: FictoshopWorld) { return world.state.products.secondary; }

Given("the catalog contains an available product", async function (this: FictoshopWorld) {
  const reference = randomUUID().slice(0, 8);
  this.state.products.primary = await this.data.createProduct(`Playwright product ${reference}`, "Product created for TypeScript cart scenarios.", 49.95, 7);
  this.state.products.secondary = await this.data.createProduct(`Second cart product ${reference}`, "Second product for TypeScript cart scenarios.", 15.25, 8);
});

for (const step of ["the floating cart is not visible", "the floating cart is hidden"]) {
  Then(step, async function (this: FictoshopWorld) { await cart(this).expectHidden(); });
}

When("the customer adds 1 available product to the cart", async function (this: FictoshopWorld) { await cart(this).addProduct(primary(this).name); });
Then("the floating cart is visible", async function (this: FictoshopWorld) { await cart(this).expectVisible(); });
Then("it shows the product name", async function (this: FictoshopWorld) { await cart(this).expectProductInCart(primary(this).name); });
Then("it shows 1 total item", async function (this: FictoshopWorld) { await cart(this).expectTotalItems(1); });
Then("it shows the correct total price", async function (this: FictoshopWorld) { await cart(this).expectGrandTotal(primary(this).price); });
Given("the customer has added 1 product to the cart", async function (this: FictoshopWorld) { await cart(this).addProduct(primary(this).name); });
When("the customer scrolls to another part of the storefront", async function (this: FictoshopWorld) { await cart(this).scroll(); });
Then("the floating cart remains inside the viewport", async function (this: FictoshopWorld) { await cart(this).expectInsideViewport(); });
When("the customer selects quantity 3", async function (this: FictoshopWorld) { await cart(this).selectQuantity(primary(this).name, 3); });
When("adds the product to the cart", async function (this: FictoshopWorld) { await cart(this).addSelectedProduct(primary(this).name); });

for (const step of ["the cart line shows quantity 3", "the line quantity is 3"]) {
  Then(step, async function (this: FictoshopWorld) { await cart(this).expectLineQuantity(primary(this).name, 3); });
}

Then("the total equals three times the unit price", async function (this: FictoshopWorld) { await cart(this).expectGrandTotal(primary(this).price * 3); });
When("the customer adds 2 more of the same product", async function (this: FictoshopWorld) { await cart(this).addProduct(primary(this).name, 2); });
Then("the cart contains one line for that product", async function (this: FictoshopWorld) { await cart(this).expectOneLine(primary(this).name); });

async function addMultiple(world: FictoshopWorld): Promise<void> {
  await cart(world).addProduct(primary(world).name, 1);
  await cart(world).addProduct(secondary(world).name, 2);
}

for (const step of ["the customer adds multiple different products", "the cart contains two different products", "the cart contains products"]) {
  Given(step, async function (this: FictoshopWorld) { await addMultiple(this); });
}

Then("every selected product is shown in the cart", async function (this: FictoshopWorld) {
  await cart(this).expectProductInCart(primary(this).name);
  await cart(this).expectProductInCart(secondary(this).name);
});
Then("total items equal the sum of all quantities", async function (this: FictoshopWorld) { await cart(this).expectTotalItems(3); });
Then("the grand total equals the sum of all line totals", async function (this: FictoshopWorld) { await cart(this).expectGrandTotal(primary(this).price + secondary(this).price * 2); });
When("the customer removes one product", async function (this: FictoshopWorld) { await cart(this).removeProduct(primary(this).name); });
Then("only that product disappears from the cart", async function (this: FictoshopWorld) {
  await cart(this).expectProductNotInCart(primary(this).name);
  await cart(this).expectProductInCart(secondary(this).name);
});
Then("the totals are recalculated", async function (this: FictoshopWorld) {
  await cart(this).expectTotalItems(2);
  await cart(this).expectGrandTotal(secondary(this).price * 2);
});
When("the customer selects \"Clear\"", async function (this: FictoshopWorld) { await cart(this).clear(); });
Then("all cart lines are removed", async function (this: FictoshopWorld) { assert.equal(await cart(this).cartItems.count(), 0); });
When("the customer reloads the storefront", async function (this: FictoshopWorld) { await cart(this).reload(); });
Then("the same product and quantity remain in the cart", async function (this: FictoshopWorld) {
  await cart(this).expectProductInCart(primary(this).name);
  await cart(this).expectLineQuantity(primary(this).name, 1);
});
When("the customer attempts to add more units than available", async function (this: FictoshopWorld) {
  const page = cart(this);
  await page.attemptOverStock(primary(this).name, primary(this).in_stock);
  this.state.lastBrowserStatus = page.lastCartResponseStatus;
});
Then("the request is rejected", function (this: FictoshopWorld) { assert.equal(this.state.lastBrowserStatus, 400); });
Then("a stock error is displayed", async function (this: FictoshopWorld) { await cart(this).expectStockError(primary(this).in_stock, primary(this).name); });
Then("the cart quantity is unchanged", async function (this: FictoshopWorld) { await cart(this).expectTotalItems(0); });
When("the customer removes that product from the cart", async function (this: FictoshopWorld) { await cart(this).removeProduct(primary(this).name); });
Then("the catalog shows the original available stock", async function (this: FictoshopWorld) { await cart(this).expectCatalogStock(primary(this).name, primary(this).in_stock); });
